File: internal/recorder/recorder.go
// Package recorder implements the out-of-band Flight Recorder: while the
// LLM payloads stay pruned and cheap, the full forensic trail (prompts,
// raw responses, routing decisions, rejected alternatives) is written to a
// local content-addressable store for human debugging.
//
// Blobs are SHA-256 content-addressed (Git-plumbing style) so repeated
// contexts deduplicate to a single object on disk.
package recorder

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// Recorder writes trace events under a base directory.
type Recorder struct {
	base string
}

// DefaultDir returns the canonical trace directory.
func DefaultDir() string {
	if p := os.Getenv("TOKENOS_TRACES"); p != "" {
		return p
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return ".tokenos-traces"
	}
	return filepath.Join(home, ".local", "state", "tokenos", "traces")
}

// New creates a Recorder rooted at dir (empty = DefaultDir).
func New(dir string) (*Recorder, error) {
	if dir == "" {
		dir = DefaultDir()
	}
	if err := os.MkdirAll(filepath.Join(dir, "objects"), 0o755); err != nil {
		return nil, err
	}
	return &Recorder{base: dir}, nil
}

// Event is one flight-recorder entry.
type Event struct {
	TaskID    string    `json:"task_id"`
	Kind      string    `json:"kind"` // decision | prompt | response | error | verify
	Summary   string    `json:"summary,omitempty"`
	BlobSHA   string    `json:"blob_sha,omitempty"`
	Timestamp time.Time `json:"ts"`
}

// putBlob stores content-addressed payload bytes, returning the SHA-256 hex.
// Identical payloads write once (deduplication).
func (r *Recorder) putBlob(data []byte) (string, error) {
	sum := sha256.Sum256(data)
	sha := hex.EncodeToString(sum[:])
	dir := filepath.Join(r.base, "objects", sha[:2])
	path := filepath.Join(dir, sha[2:])
	if _, err := os.Stat(path); err == nil {
		return sha, nil // already stored
	}
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return "", err
	}
	return sha, os.WriteFile(path, data, 0o600)
}

// Record writes a full payload blob plus an index line into the per-task
// journal (NDJSON, append-only).
func (r *Recorder) Record(taskID, kind, summary string, payload []byte) (string, error) {
	sha := ""
	if len(payload) > 0 {
		var err error
		sha, err = r.putBlob(payload)
		if err != nil {
			return "", err
		}
	}
	ev := Event{TaskID: taskID, Kind: kind, Summary: summary, BlobSHA: sha, Timestamp: time.Now().UTC()}
	line, err := json.Marshal(ev)
	if err != nil {
		return "", err
	}
	journal := filepath.Join(r.base, fmt.Sprintf("%s.ndjson", sanitize(taskID)))
	f, err := os.OpenFile(journal, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o600)
	if err != nil {
		return "", err
	}
	defer f.Close()
	if _, err := f.Write(append(line, '\n')); err != nil {
		return "", err
	}
	return sha, nil
}

// Events replays the journal for a task.
func (r *Recorder) Events(taskID string) ([]Event, error) {
	journal := filepath.Join(r.base, fmt.Sprintf("%s.ndjson", sanitize(taskID)))
	data, err := os.ReadFile(journal)
	if os.IsNotExist(err) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	var out []Event
	start := 0
	for i := 0; i <= len(data); i++ {
		if i == len(data) || data[i] == '\n' {
			if i > start {
				var ev Event
				if json.Unmarshal(data[start:i], &ev) == nil {
					out = append(out, ev)
				}
			}
			start = i + 1
		}
	}
	return out, nil
}

// Blob fetches a stored payload by SHA.
func (r *Recorder) Blob(sha string) ([]byte, error) {
	if len(sha) < 3 {
		return nil, fmt.Errorf("invalid sha")
	}
	return os.ReadFile(filepath.Join(r.base, "objects", sha[:2], sha[2:]))
}

func sanitize(s string) string {
	out := make([]rune, 0, len(s))
	for _, r := range s {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9', r == '-', r == '_':
			out = append(out, r)
		default:
			out = append(out, '_')
		}
	}
	return string(out)
}

File: src/recorder.rs
//! Flight Recorder: out-of-band forensic trail. While the LLM payloads stay
//! pruned and cheap, the full trail (prompts, raw responses, routing
//! decisions) is written to a local content-addressable store for human
//! debugging.
//!
//! Blobs are SHA-256 content-addressed (Git-plumbing style) so repeated
//! contexts deduplicate to a single object on disk.

use anyhow::{anyhow, Result};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};

/// One flight-recorder entry.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Event {
    pub task_id: String,
    /// decision | prompt | response | error | verify
    pub kind: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub summary: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub blob_sha: String,
    pub ts: DateTime<Utc>,
}

/// Recorder writes trace events under a base directory.
pub struct Recorder {
    base: PathBuf,
}

/// Canonical trace directory: $TOKENOS_TRACES or ~/.local/state/tokenos/traces.
pub fn default_dir() -> PathBuf {
    if let Ok(p) = std::env::var("TOKENOS_TRACES") {
        if !p.is_empty() {
            return PathBuf::from(p);
        }
    }
    match dirs::home_dir() {
        Some(home) => home.join(".local").join("state").join("tokenos").join("traces"),
        None => PathBuf::from(".tokenos-traces"),
    }
}

impl Recorder {
    /// Create a Recorder rooted at dir (None = default_dir()).
    pub fn new(dir: Option<&Path>) -> Result<Self> {
        let base = match dir {
            Some(d) if !d.as_os_str().is_empty() => d.to_path_buf(),
            _ => default_dir(),
        };
        fs::create_dir_all(base.join("objects"))?;
        Ok(Self { base })
    }

    /// Store content-addressed payload bytes, returning the SHA-256 hex.
    /// Identical payloads write once (deduplication).
    fn put_blob(&self, data: &[u8]) -> Result<String> {
        let sha = hex::encode(Sha256::digest(data));
        let dir = self.base.join("objects").join(&sha[..2]);
        let path = dir.join(&sha[2..]);
        if path.exists() {
            return Ok(sha); // already stored
        }
        fs::create_dir_all(&dir)?;
        fs::write(&path, data)?;
        Ok(sha)
    }

    /// Write a full payload blob plus an index line into the per-task
    /// journal (NDJSON, append-only). Returns the blob SHA (empty if no payload).
    pub fn record(&self, task_id: &str, kind: &str, summary: &str, payload: &[u8]) -> Result<String> {
        let sha = if payload.is_empty() {
            String::new()
        } else {
            self.put_blob(payload)?
        };
        let ev = Event {
            task_id: task_id.to_string(),
            kind: kind.to_string(),
            summary: summary.to_string(),
            blob_sha: sha.clone(),
            ts: Utc::now(),
        };
        let line = serde_json::to_string(&ev)?;
        let journal = self.base.join(format!("{}.ndjson", sanitize(task_id)));
        let mut f = fs::OpenOptions::new().create(true).append(true).open(&journal)?;
        f.write_all(line.as_bytes())?;
        f.write_all(b"\n")?;
        Ok(sha)
    }

    /// Replay the journal for a task.
    pub fn events(&self, task_id: &str) -> Result<Vec<Event>> {
        let journal = self.base.join(format!("{}.ndjson", sanitize(task_id)));
        let data = match fs::read_to_string(&journal) {
            Ok(d) => d,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(Vec::new()),
            Err(e) => return Err(e.into()),
        };
        let mut out = Vec::new();
        for line in data.lines() {
            if line.trim().is_empty() {
                continue;
            }
            if let Ok(ev) = serde_json::from_str::<Event>(line) {
                out.push(ev);
            }
        }
        Ok(out)
    }

    /// Fetch a stored payload by SHA.
    pub fn blob(&self, sha: &str) -> Result<Vec<u8>> {
        if sha.len() < 3 {
            return Err(anyhow!("invalid sha"));
        }
        Ok(fs::read(self.base.join("objects").join(&sha[..2]).join(&sha[2..]))?)
    }
}

fn sanitize(s: &str) -> String {
    s.chars()
        .map(|c| match c {
            'a'..='z' | 'A'..='Z' | '0'..='9' | '-' | '_' => c,
            _ => '_',
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn temp_recorder() -> (Recorder, tempdir::TempDirGuard) {
        let dir = tempdir::make();
        let r = Recorder::new(Some(dir.path())).unwrap();
        (r, dir)
    }

    // Minimal self-contained temp-dir helper (avoids extra dev-dependency).
    mod tempdir {
        use std::path::{Path, PathBuf};
        pub struct TempDirGuard(PathBuf);
        impl TempDirGuard {
            pub fn path(&self) -> &Path {
                &self.0
            }
        }
        impl Drop for TempDirGuard {
            fn drop(&mut self) {
                let _ = std::fs::remove_dir_all(&self.0);
            }
        }
        pub fn make() -> TempDirGuard {
            let p = std::env::temp_dir().join(format!(
                "tokenos-rec-test-{}-{}",
                std::process::id(),
                std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_nanos()
            ));
            std::fs::create_dir_all(&p).unwrap();
            TempDirGuard(p)
        }
    }

    #[test]
    fn record_and_replay() {
        let (r, _g) = temp_recorder();
        let sha = r.record("task-1", "prompt", "the prompt", b"PAYLOAD BYTES").unwrap();
        assert_eq!(sha.len(), 64);
        r.record("task-1", "response", "the answer", b"RESPONSE").unwrap();
        let evs = r.events("task-1").unwrap();
        assert_eq!(evs.len(), 2);
        assert_eq!(evs[0].kind, "prompt");
        assert_eq!(evs[1].kind, "response");
        let blob = r.blob(&sha).unwrap();
        assert_eq!(blob, b"PAYLOAD BYTES");
    }

    #[test]
    fn blob_deduplication() {
        let (r, _g) = temp_recorder();
        let a = r.record("t", "prompt", "", b"same content").unwrap();
        let b = r.record("t", "prompt", "", b"same content").unwrap();
        assert_eq!(a, b);
    }

    #[test]
    fn empty_payload_has_no_blob() {
        let (r, _g) = temp_recorder();
        let sha = r.record("t", "decision", "route=DIRECT", b"").unwrap();
        assert!(sha.is_empty());
        let evs = r.events("t").unwrap();
        assert_eq!(evs.len(), 1);
        assert!(evs[0].blob_sha.is_empty());
    }

    #[test]
    fn missing_journal_returns_empty() {
        let (r, _g) = temp_recorder();
        assert!(r.events("nope").unwrap().is_empty());
    }

    #[test]
    fn sanitize_path_traversal() {
        assert_eq!(sanitize("../../etc/passwd"), "______etc_passwd");
        assert_eq!(sanitize("task-42_ok"), "task-42_ok");
    }

    #[test]
    fn invalid_sha_rejected() {
        let (r, _g) = temp_recorder();
        assert!(r.blob("ab").is_err());
    }
}
